CONTENT: two new blogs on package health (#116)
pyOpenSci/pyopensci.github.io: _posts/2022-11-10-why-python-package-health-metrics-matter.md

---
layout: single
title: "Why Thinking About Metrics Should be the First Thing That Any New Organization, Program or Project Does"
excerpt: "Collecting baseline data that aligns with the goals and outcomes of your project, program or organization is critical to do at the beginning. Here I briefly explain why Python package health metrics are so important to the long-term success of pyOpenSci."
author: "Leah Wasser"
permalink: /blog/why-python-package-health-metrics-matter
header:
    overlay_color: "#666"
    overlay_filter: 0.6
categories:
  - blog-post
  - highlight
  - python-packaging
  - peer-review
toc: true
comments: true
---

> 

Reviewer: Should there be a brief 1-2 sentence intro here with a link back to the previous post about "why open source package health matters to scientists"?

I actually like way it starts now with the assertive "This isn't my first rodeo" but as I understand it now you want this to be a series of posts and I totally didn't get that on the first read. So intro could be short and sweet, like "In [my last post](link) I talked about why OSS package health matters to scientists. Now I want to talk about why it matters for pyOpenSci..."

Reviewer: I do see now you say "in a previous post" below, linking back to "why OSS scientists". Maybe I'm just a bad reader/listener, and obviously you have your own voice, but I might be a bit more assertive about linking the posts right at the top

Author: ok will look at that. yes this is a three part series. so i do want this to be clear. i originally wrote it as one ginormous and then realized it was way too long with too many focus points.

## Metrics are critical to the development of any program 

I've created a few open science focused programs now from the ground up. One at 
NEON and another at CU Boulder. When building a new program, one of the first 
things that I do (after defining the mission and goals) is to define the metrics 
that constitute success. 

These metrics are critical to define early because:

> * They drive everything that you do

Reviewer: ```suggestion
* They drive everything that you do.
```
Suggest either making all of these bullet points sentences with periods or none of them, but not mixing

3. How/if we contributed to improving the health of packages in our ecosystem
4. How we are impacting the broader scientific python, open source community

### Baseline metrics collected from the start will help your future self when running or working in an organization

As mentioned above, collecting metrics from the start 
of your efforts allows you to get off the ground running with data that you can use to compare to future data. Thus while it may not be the work that you want to do, it will 
help your future self. 

For pyOpenSci, collecting metrics allows us in the future 
to evaluate our programs and adaptively change things to 
make sure we are getting the outcomes that we want. 

Outcomes such as 
* Scientists being better able to find packages that are maintained to support their workflows  
* Improved documentation in packages as a result of our reviews

### But what metrics should we collect about scientific Python packages?

[In a previous post,](/blog/2022-10-24-why-should-python-open-source-matter-science) I spoke 
generally about why open source should matter to you as a scientist and as a 
developer or package maintainer. 

> To better understand what data we should be collecting to track our packages 

Reviewer: ```suggestion
To better understand what data we should be collecting to track our packages' 
```
missing possessive apostrophe, "packages' heatlh"

health over time, I went to Twitter to see what my colleagues around the world
had to say. That conversation resulted in some really interesting insights.

> 

Reviewer: Maybe state briefly here "I will give some example from this conversation in my next post, the last in this series, but first I want to sum up where I'm at after that discussion"

Author: ok maybe that will help with some of the comments above?

Most importantly, it allowed me to begin to break down and group metrics in 
terms of pyOpenSci goals. 

> 

Reviewer: Maybe include some example replies here from the twitter conversation to show how they guided your thinking?

Reviewer: ah nvm, I see that's the next post. Maybe state briefly here? "I will dive into these in the next post, the last in this series, but first I want to sum up where I'm at after that discussion"

### Goals for package metrics

We hope that:

* Peer review improves Python package structure and usability. 
* Peer review in some way supports maintenance and/or responsible archiving when a package comes to life-end.

> * Over time, the package is improved and maintained with possible contributions for those other than the maintainer. 

Reviewer: Not quite sure what you want to say here, "with possible contributions for those other than the maintainer"

Something like "improved and maintained in such a way that it's possible for others to contribute besides than the maintainer"?

Author: i'm going to commit small edits first and will go back and work on these comments locally next!

We need metrics to understand things like

* Community adoption of the package (are scientists using it?)
* Maintenance level of the package (are maintainers still working on it and fixing bugs?)

> * Infrastructure (are tests setup to help identify if contributions break things? )

Reviewer: ```suggestion
* Infrastructure (are tests set up to help identify if contributions break things? )
```
Still reading overall (sounds great far!) but just a very minor suggested edit.

I am pretty sure that when using "set up" as a phrasal verb and not a noun, you want a space between the two words. 

It's confusing here because of the copula "are", that "set up" is modifying as a helper, kind of like a gerund, "are test running? tests are running". 

You would write "Tests are set up to", not "test are setup to"; you wouldn't write "I am setup to fail"

I don't find any good prescriptive grammarian blog posts to tell me if I'm right, but

"are set up to" returns [43.8 M results](https://www.google.com/search?q=%22are+set+up+to%22&sxsrf=ALiCzsbLRsvFrbJkVkkTWXzPpg_wKFLv6Q%3A1667495747932&source=hp&ei=Q_djY4LYNdTZ9API-7OoCQ&iflsig=AJiK0e8AAAAAY2QFUwrbMO9LLNMI0lO9J2pmgLgQFWg1&ved=0ahUKEwjChd7dwZL7AhXULH0KHcj9DJUQ4dUDCAk&uact=5&oq=%22are+set+up+to%22&gs_lcp=Cgdnd3Mtd2l6EAMyBggAEBYQHjIGCAAQFhAeMggIABAWEB4QDzIGCAAQFhAeMgUIABCGAzIFCAAQhgM6BAgjECc6BQgAEJECOggILhDUAhCRAjoRCC4QgwEQxwEQsQMQ0QMQgAQ6CwgAEIAEELEDEIMBOg4ILhCABBCxAxDHARDRAzoECAAQQzoICAAQgAQQsQM6CggAELEDEIMBEEM6BQgAEIAEOgcIABCxAxBDOhEILhCABBCxAxCDARDHARDRAzoICC4QsQMQgwE6CwguEIAEEMcBEK8BOgsILhCABBCxAxCDAToICC4QgAQQsQM6CAgAELEDEIMBOgsILhCABBCxAxDUAjoICAAQsQMQkQI6BQguEIAEOggIABAWEB4QClAAWI0ZYPgaaAZwAHgAgAG8AYgB1hSSAQQxLjIwmAEAoAEB&sclient=gws-wiz)
"are setup to" return [195k results](https://www.google.com/search?q=%22are+setup+to%22&sxsrf=ALiCzsblv7LzR7ffOBefwkolNZHe20I1mg%3A1667495752615&ei=SPdjY5eLJfCHptQP0bSNuAM&ved=0ahUKEwiXz_7fwZL7AhXwg4kEHVFaAzcQ4dUDCA8&uact=5&oq=%22are+setup+to%22&gs_lcp=Cgxnd3Mtd2l6LXNlcnAQAzIGCAAQBxAeMgoIABAIEAcQHhAPMgUIABCGAzIFCAAQhgMyBQgAEIYDOgoIABBHENYEELADSgQIQRgASgQIRhgAUPgEWPgEYO0GaAFwAXgAgAGcAYgBnAGSAQMwLjGYAQCgAQHIAQXAAQE&sclient=gws-wiz-serp)

* Usability (is the package documented in a way that helps users quickly get started)

> ## Three Python open source software healthy metric "buckets" that pyOpenSci cares about

Reviewer: ```suggestion
## Three metric buckets that pyOpenSci cares about
```
This noun stack is very hard for my easily distracted brain to read: "Three Python open source software healthy metric 'buckets'"

> Based on all of the feedback, and what I *think* might be a start at what pyOpenSci  

Reviewer: Overall everything you're saying here makes perfect sense to me, but I was kind of surprised this was the end of the post. I thought you were going to elaborate more on why you chose those 3 buckets, e.g., based on comments on twitter. Right now it sounds a bit like you are repeating what you just said above, and that you are defining your metrics to be exactly the things you just said we need to understand? Maybe give a little more detail about how you would actually compute each metric here?
For example "to look at community adoption we will consider number of github stars, number of downloads and forks, and number of stackoverflow posts" (or whatever). This could be a good place to reference rOpenSci and fatiando as an illustration of how other communities have approached this?

I know you're trying to close out the post here and I'm not trying to suggest you add a ton more. Maybe merge this section and the one above it on goals? "Goals of metrics and three possible buckets"

Author: totally. this blog was supposed to setup the next blog and that summary at the end is supposed to kind of setup - if you want to understand where these came from - read the next blog. but it sounds like that wasn't clear in the post? 
In terms of how we calculate i was thinking that would be a 4th blog.

let me know what you think as i don't want to leave people hanging here.

Reviewer: Overall the post reads clearly to me now but it still feels a bit like you repeat yourself at the end, and now that I have read the first post, it feels like you are repeating yourself from that post too

If it were me I would rewrite like so:
* in my last post I translated what scientists need from open source packages into some goals for package metrics, [here](https://www.pyopensci.org/blog/why-python-open-source-software-matters-for-scientists#goals-for-package-metrics). Let me restate those here:
  * goals, leave out the metrics since you're about to say this anyways
* We need metrics that match these goals, but as I discussed this with others I realized that we also need a holistic view of package health.
* I have chosen three dimensions (or however many you decide on) along which we can consider package health, and we can group any number of metrics along these dimensions
  * maintenance
  * infrastructure
  * community
In the last post in this series I will share some of the discussion on twitter that helped me come to this conclusion. I hope you'll continue the discussion here and on twitter as well.
